### PR TITLE
Update Connectivity Test Readme

### PR DESCRIPTION
**Pull request by yophilav**
- Update connectivity test readme with the following cases:
1. Cloud to Module Direct Method Test
2. Cloud to EdgeAgent Direct Method Test (PING)
3. Cloud to Device Message Test
4. EdgeHub Restart Test : Message Test
5. EdgeHub Restart Test : Direct Method Test

### CODE REVIEW

File: test/connectivity/README.md [hunk 1]
```diff
@@ -69,3 +72,106 @@ Network offline/online should not affect this test result since Edge hub stores
 6. When DeploymentTester2 starts up, it will get all environment variables and report to Test Result Coordinator (#5).
 
 When network goes offline, route#3 will not happen.  But once the network comes back online, Edge Agent should be able to get the latest updated deployment and deploy DeploymentTester2.
+
+#### Cloud to Module Direct Method Test
+
+![Connectivity Test Diagram](./images/ConnectivityTest_C2M_DirectMethod.png)
+
+The Direct Method Connectivity Test is designed to test direct method resilience against network connectivity issue. 
+##### Test Specific Component:
```
Comment (philipktlin) [inline test/connectivity/README.md]: It is good to have standard sections - Test specific component and test mechanism/flow.  Please create a task to update Messaging test and deployment test to follow.

File: test/connectivity/README.md [hunk 2]
```diff
@@ -69,3 +72,106 @@ Network offline/online should not affect this test result since Edge hub stores
...
+###### Test Mechanism: 
+	1. The EdgeHub Restart Tester issue a direct method to EdgeAgent notifying the EdgeAgent to restart EdgeHub module
+	2. Once the restart direct method response with a success, the EdgeHub Restart Tester attempts to send a message to Message Receiver via EdgeHub. 
+		a. The SDK will throw an exception within EdgeHub Restart Tester if the EdgeHub is unabled to be connected
+		b. If the SDK stops throwing an exception and respond with HTTP OK, it is an indicator that the sender module is able to communicate with EdgeHub. However, this does not imply that the EdgeHub can successfully send the message to the Message Receiver.
+	3. The EdgeHub Restart Tester reports its { restart time, message sending response, and message sending completion time } to the Test Result Coordinator
+	4. Eventually the EdgeHub will forward the message to the Message Receiver once the receiver module can successfully reconnected to the EdgeHub
+	5. Once the receiver module receives a message, it will report the message receival to Test Result Coordinator
+
+###### Passing Criteria:
```
Comment (philipktlin) [inline test/connectivity/README.md]: should we add passing criteria for every test?

Comment (yophilav) [reply test/connectivity/README.md]: I think it makes sense to have a passing criteria for test cases, otherwise, the test case doesn't mean anything.

Comment (philipktlin) [reply test/connectivity/README.md]: Please add passing criteria for Cloud to EdgeAgent Direct Method Test (PING) and Cloud to Module Direct Method Test.  thanks.

File: test/connectivity/README.md [hunk 3]
```diff
@@ -69,3 +72,106 @@ Network offline/online should not affect this test result since Edge hub stores
...
 
 When network goes offline, route#3 will not happen.  But once the network comes back online, Edge Agent should be able to get the latest updated deployment and deploy DeploymentTester2.
+
+#### Cloud to Module Direct Method Test
+
+![Connectivity Test Diagram](./images/ConnectivityTest_C2M_DirectMethod.png)
+
+The Direct Method Connectivity Test is designed to test direct method resilience against network connectivity issue. 
+##### Test Specific Component:
+	1. Direct Method (Cloud) Sender is an Azure service client that is wrapped in a container module that is deployed into a test agent.  It is responsible for sending a "HelloWorldMethod" direct method to the Direct Method Receiver module and report the direct method status to Test Result Coordinator
```
Comment (philipktlin) [inline test/connectivity/README.md]: add period at the end.

File: test/connectivity/README.md [hunk 4]
```diff
@@ -69,3 +72,106 @@ Network offline/online should not affect this test result since Edge hub stores
...
+	2. Direct Method Receiver is a module container that runs a module client that is deployed into a test agent. It is a targeted receiver of Direct Method Sender. It is responsible for replying to the "HelloWorldMethod" direct method and notify Test Result Coordinate if it receives a direct method.
```
Comment (philipktlin) [inline test/connectivity/README.md]:
```suggestion
	2. Direct Method Receiver is a module container that runs a module client that is deployed into a test agent. It is a targeted receiver of Direct Method (Cloud) Sender. It is responsible for replying to the "HelloWorldMethod" direct method and notify Test Result Coordinate if it receives a direct method.
```

File: test/connectivity/README.md (hunk 3 above)
Comment (philipktlin) [reply test/connectivity/README.md]: DM Cloud sender is in host network, do you think we need to explicit mention it here.

File: test/connectivity/README.md [hunk 5]
```diff
@@ -69,3 +72,106 @@ Network offline/online should not affect this test result since Edge hub stores
...
+
+##### Test Mechanism: 
+	1. The Direct Method Cloud Sender invokes a test direct method to the IoT Hub
```
Comment (philipktlin) [inline test/connectivity/README.md]: add period at the end for each bullet point

File: test/connectivity/README.md [hunk 6]
```diff
@@ -69,3 +72,106 @@ Network offline/online should not affect this test result since Edge hub stores
...
+	2. The IoT Hub forwards the direct method to Edge Hub
+	3. The Edge Hub forwards the direct method to Direct Method Receiver
+	4. The Direct Method Receiver notifies Test Analyzer regarding the received direct method
```
Comment (philipktlin) [inline test/connectivity/README.md]:
```suggestion
	4. The Direct Method Receiver notifies test result coordinator regarding the received direct method
```

File: test/connectivity/README.md [hunk 7]
```diff
@@ -69,3 +72,106 @@ Network offline/online should not affect this test result since Edge hub stores
...
+	5. The Direct Method Receiver responds to Edge Hub's direct method invocation
+	6. The Edge Hub  responds to IoT Hub direct method invocation
```
Comment (philipktlin) [inline test/connectivity/README.md]: remove extra space
```suggestion
	6. The Edge Hub responds to IoT Hub direct method invocation 
```

File: test/connectivity/README.md [hunk 8]
```diff
@@ -69,3 +72,106 @@ Network offline/online should not affect this test result since Edge hub stores
...
+	2. Direct Method Receiver is a module container that runs a module client that is deployed into a test agent. It is a targeted receiver of Direct Method Sender. It is responsible for replying to the "HelloWorldMethod" direct method and notify Test Result Coordinate if it receives a direct method.
+
+##### Test Mechanism: 
+	1. The Direct Method Cloud Sender invokes a test direct method to the IoT Hub
+	2. The IoT Hub forwards the direct method to Edge Hub
+	3. The Edge Hub forwards the direct method to Direct Method Receiver
+	4. The Direct Method Receiver notifies Test Analyzer regarding the received direct method
+	5. The Direct Method Receiver responds to Edge Hub's direct method invocation
+	6. The Edge Hub  responds to IoT Hub direct method invocation 
+	7. The Direct Method Cloud Sender receives a response from IoT Hub
```
Comment (philipktlin) [inline test/connectivity/README.md]:
```suggestion
	7. The Direct Method (Cloud) Sender receives a response from IoT Hub
```

File: test/connectivity/README.md [hunk 9]
```diff
@@ -69,3 +72,106 @@ Network offline/online should not affect this test result since Edge hub stores
...
+	8. The Direct Method Cloud Sender notifies Test Analyzer regarding its direct method response
```
Comment (philipktlin) [inline test/connectivity/README.md]:
```suggestion
	8. The Direct Method Cloud Sender notifies test result coordinator regarding its direct method response
```

File: test/connectivity/README.md [hunk 10]
```diff
@@ -69,3 +72,106 @@ Network offline/online should not affect this test result since Edge hub stores
...
+Test Analyzer compile a report from Direct Method Cloud Sender, Direct Method Receiver, Network Controller and determine if the test case is passed or failed.
```
Comment (philipktlin) [inline test/connectivity/README.md]: Please globally replace test analyzer by test result coordinator.

File: test/connectivity/README.md [hunk 11]
```diff
@@ -69,3 +72,106 @@ Network offline/online should not affect this test result since Edge hub stores
...
+	3. The Edge Agent has a pre-implemented "ping" method with a response.
+	4. The Direct Method Cloud Sender receives a response from IoT Hub
+	5. The Direct Method Cloud Sender notifies Test Analyzer regarding its direct method response
+Test Analyzer compile a report from Direct Method Cloud Sender, Direct Method Receiver, Network Controller and determine if the test case is passed or failed. 
+
+#### Cloud to Device Message Test
+
+![Connectivity Test Diagram](./images/ConnectivityTest_C2D_Message.png)
+
+The goal of the test scheme is to verify the recoverability of the cloud to device messaging once the device goes offline.
```
Comment (philipktlin) [inline test/connectivity/README.md]: should we add a section called "Test goal" for all tests?

File: test/connectivity/README.md [hunk 12]
```diff
@@ -69,3 +72,106 @@ Network offline/online should not affect this test result since Edge hub stores
...
+	5. The Direct Method Cloud Sender notifies Test Analyzer regarding its direct method response
+Test Analyzer compile a report from Direct Method Cloud Sender, Direct Method Receiver, Network Controller and determine if the test case is passed or failed. 
+
+#### Cloud to Device Message Test
+
+![Connectivity Test Diagram](./images/ConnectivityTest_C2D_Message.png)
+
+The goal of the test scheme is to verify the recoverability of the cloud to device messaging once the device goes offline. 
+##### Test Specific Components:
+	1. Message Sender is a Azure service client that is deployed into a test agent as a container module. The responsibility of the module is to use the Azure API to trigger IoT Hub to send a message to device along with reporting to Test Result Coordinator once a message is sent.
```
Comment (philipktlin) [inline test/connectivity/README.md]:
```suggestion
	1. Message Sender is an Azure service client that is deployed into a test agent as a container module. The responsibility of the module is to use the Azure API to trigger IoT Hub to send a message to device along with reporting to Test Result Coordinator once a message is sent. 
```

File: test/connectivity/README.md [hunk 13]
```diff
@@ -69,3 +72,106 @@ Network offline/online should not affect this test result since Edge hub stores
...
+Test Analyzer compile a report from Direct Method Cloud Sender, Direct Method Receiver, Network Controller and determine if the test case is passed or failed. 
+
+#### Cloud to Device Message Test
+
+![Connectivity Test Diagram](./images/ConnectivityTest_C2D_Message.png)
+
+The goal of the test scheme is to verify the recoverability of the cloud to device messaging once the device goes offline. 
+##### Test Specific Components:
+	1. Message Sender is a Azure service client that is deployed into a test agent as a container module. The responsibility of the module is to use the Azure API to trigger IoT Hub to send a message to device along with reporting to Test Result Coordinator once a message is sent. 
+	2. Message Receiver is a device client that is wrapped in a module container for a deployment into a test agent. The device module behaves as a target receiving end of the message sending from Message Sender and is responsible for reporting to the Test Result Coordinator once a message is received.
```
Comment (philipktlin) [inline test/connectivity/README.md]:
```suggestion
	2. Message Receiver is a device client that is wrapped in a module container for a deployment into a test agent. The device module behaves as a target receiver of the message sending from Message Sender and is responsible for reporting to the Test Result Coordinator once a message is received.
```

File: test/connectivity/README.md [hunk 14]
```diff
@@ -69,3 +72,106 @@ Network offline/online should not affect this test result since Edge hub stores
...
+
+##### Test Implementation Assumption:
```
Comment (philipktlin) [inline test/connectivity/README.md]: suggest to simplify
```suggestion
##### Test Assumptions:
```

File: test/connectivity/README.md [hunk 15]
```diff
@@ -69,3 +72,106 @@ Network offline/online should not affect this test result since Edge hub stores
...
+	1. The cloud takes less than 15 minutes to forward all of messages from the cloud to device when the Edge Hub is online.
+	2. Each cycle of Network Controller online-offline period must be longer than the Edge Hub recovering period
```
Comment (philipktlin) [inline test/connectivity/README.md]: what happens if network online-offline period is shorter than Edge Hub recovering period?

File: test/connectivity/README.md [hunk 16]
```diff
@@ -69,3 +72,106 @@ Network offline/online should not affect this test result since Edge hub stores
...
+	3. Since the Message Sender uses service client to send a message to the cloud from the host network, it will always report success even if the message has yet gotten to the Message Receiver. If the Message Sender fails to send a message, the message sequencing ID should NOT be increased; instead, a failed message with the same sequence number should be resent.
```
Comment (philipktlin) [inline test/connectivity/README.md]: Please mention why and the error still logged to Message Sender module log.

File: test/connectivity/README.md [hunk 17]
```diff
@@ -69,3 +72,106 @@ Network offline/online should not affect this test result since Edge hub stores
...
+	
+##### Test Mechanism: 
+	1. The Message Sender invokes the IoT Hub to send  a message to a targeted device
```
Comment (philipktlin) [inline test/connectivity/README.md]:
```suggestion
	1. The Message Sender invokes the IoT Hub to send a message to a targeted device
```

File: test/connectivity/README.md [hunk 18]
```diff
@@ -69,3 +72,106 @@ Network offline/online should not affect this test result since Edge hub stores
...
+	2. The Message Sender notify the Test Result Coordinator that a message has been sent.
+	3. The IoT Hub sends a message to Edge Hub addressing a targeted device
+	4. The Edge Hub forward the message to a Message Receiver device module.
+	5. The Message Receiver notifies that the message successfully received to the Test Result Coordinator
+	6. The Message Receiver replies that the message successfully received to the Edge Hub
+	7. The Edge Hub forward the Message Receiver's response back to IoT Hub
+	8. The IoT Hub forward the response to Message Sender.
+
+##### Passing Criteria: 
+All the messages sent by Message Sender is received in order at Message Receiver within (EdgeHub_Recover_Time) + 15 minutes.
```
Comment (philipktlin) [inline test/connectivity/README.md]: what is edgehub_recover_time?

File: test/connectivity/README.md [hunk 19]
```diff
@@ -69,3 +72,106 @@ Network offline/online should not affect this test result since Edge hub stores
...
+
+#### EdgeHub Restart Test
+The goal of EdgeHub Restart Test is to verify/benchmark a restart period of an EdgeHub. The measurement is done using two methods: a Direct Method and a Message.
+
+The message test is developed to strictly test only the only when the EdgeHub complete its restart and is ready for a module to connect.  While the direct method test also make sure that the receiver can successfully response to the direct method request from the sender. This results in a slightly different time measurement between the two test (see the diagram below).
```
Comment (philipktlin) [inline test/connectivity/README.md]:
```suggestion
The message test is developed to strictly test only when the EdgeHub complete its restart and is ready for a module to connect.  While the direct method test also make sure that the receiver can successfully response to the direct method request from the sender. This results in a slightly different time measurement between the two test (see the diagram below).
```

File: test/connectivity/README.md [hunk 20]
```diff
@@ -69,3 +72,106 @@ Network offline/online should not affect this test result since Edge hub stores
...
+	1. The EdgeHub module can successfully be restarted by the EdgeHub Restart Tester
+	2. The EdgeHub Restart Tester can successfully send a message to Edge Hub within one minute after the EdgeHub is issued a restart
+The Message Receiver receives a message from EdgeHub Restart Tester that is routed via EdgeHub within a reasonable timely fashion
+
+##### EdgeHub Restart Test : Direct Method Test
+
+![Connectivity Test Diagram](./images/ConnectivityTest_EdgeHubRestart_DirectMethod.png)
+
+###### Test Mechanism: 
+	1. The EdgeHub Restart Tester issue a direct method to EdgeAgent notifying the EdgeAgent to restart EdgeHub module
```
Comment (philipktlin) [inline test/connectivity/README.md]:
```suggestion
	1. The EdgeHub Restart Tester issues a direct method to EdgeAgent notifying the EdgeAgent to restart EdgeHub module
```

File: test/connectivity/README.md [hunk 21]
```diff
@@ -69,3 +72,106 @@ Network offline/online should not affect this test result since Edge hub stores
...
+
+The message test is developed to strictly test only the only when the EdgeHub complete its restart and is ready for a module to connect.  While the direct method test also make sure that the receiver can successfully response to the direct method request from the sender. This results in a slightly different time measurement between the two test (see the diagram below).
+
+![Connectivity Test Diagram](./images/ConnectivityTest_EdgeHubRestart_Timeline.png)
+
+##### EdgeHub Restart Test : Message Test
+
+![Connectivity Test Diagram](./images/ConnectivityTest_EdgeHubRestart_Message.png)
+
+###### Test Mechanism:
```
Comment (philipktlin) [inline test/connectivity/README.md]: Why Test Components section is missing here? And we should mention EdgeHub Restart Tester and edge agent in host network.

File: test/connectivity/README.md [hunk 22]
```diff
@@ -69,3 +72,106 @@ Network offline/online should not affect this test result since Edge hub stores
...
+##### EdgeHub Restart Test : Message Test
+
+![Connectivity Test Diagram](./images/ConnectivityTest_EdgeHubRestart_Message.png)
+
+###### Test Mechanism: 
+	1. The EdgeHub Restart Tester issue a direct method to EdgeAgent notifying the EdgeAgent to restart EdgeHub module
+	2. Once the restart direct method response with a success, the EdgeHub Restart Tester attempts to send a message to Message Receiver via EdgeHub. 
+		a. The SDK will throw an exception within EdgeHub Restart Tester if the EdgeHub is unabled to be connected
+		b. If the SDK stops throwing an exception and respond with HTTP OK, it is an indicator that the sender module is able to communicate with EdgeHub. However, this does not imply that the EdgeHub can successfully send the message to the Message Receiver.
+	3. The EdgeHub Restart Tester reports its { restart time, message sending response, and message sending completion time } to the Test Result Coordinator
```
Comment (philipktlin) [inline test/connectivity/README.md]: too specific; may easily be forgotten to update if code changed.
```suggestion
	3. The EdgeHub Restart Tester reports restart time and message related results to the Test Result Coordinator
```

File: test/connectivity/README.md (hunk 1 above)
Comment (yophilav) [reply test/connectivity/README.md]: What PBI should I put this under?

File: test/connectivity/README.md (hunk 11 above)
Comment (yophilav) [reply test/connectivity/README.md]: I don't know. I'll leave it up to your call :/

File: test/connectivity/README.md (hunk 18 above)
Comment (yophilav) [reply test/connectivity/README.md]: It is how long edgeHub realize it is online/offline. This number can change depending on the SDK; hence, I'm not going to hard-coded it here.

File: test/connectivity/README.md (hunk 21 above)
Comment (yophilav) [reply test/connectivity/README.md]: OOOH I forgot.

File: test/connectivity/README.md (hunk 1 above)
Comment (philipktlin) [reply test/connectivity/README.md]: if not pbi fit, just create a bug which to align section format.

File: test/connectivity/README.md (hunk 11 above)
Comment (philipktlin) [reply test/connectivity/README.md]: leave it for now.

File: test/connectivity/README.md [hunk 23]
```diff
@@ -69,3 +77,130 @@ Network offline/online should not affect this test result since Edge hub stores
...
 When network goes offline, route#3 will not happen.  But once the network comes back online, Edge Agent should be able to get the latest updated deployment and deploy DeploymentTester2.
+
+#### Cloud to Module Direct Method Test
+
+![Connectivity Test Diagram](./images/ConnectivityTest_C2M_DirectMethod.png)
+
+The Direct Method Connectivity Test is designed to test direct method resilience against network connectivity issue. 
+##### Test Specific Component:
+	1. Direct Method (Cloud) Sender is an Azure service client that is wrapped in a container module that is deployed into a test agent.  It is responsible for sending a "HelloWorldMethod" direct method to the Direct Method Receiver module and report the direct method status to Test Result Coordinator.
+	2. Direct Method Receiver is a module container that runs a module client that is deployed into a test agent. It is a targeted receiver of Direct Method Sender. It is responsible for replying to the "HelloWorldMethod" direct method and notify Test Result Coordinate if it receives a direct method.
```
[the remark this answers is not in the document]
Comment (kkdawkins) [reply test/connectivity/README.md]: Otherwise, it LGTM

**Pull request merged**